fix: improved scene event type names and new events with associated delegate declarations

NoelStephensUnity opened the pull request:
Per feedback from MTT hack week, it was determined that we should rename the enums defined in SceneEventData.SceneEventTypes while also providing users with the option to subscribe to each scene event type individually while leaving the optional OnSceneEvent for users that preferred to receive all scene event type notifications in one place.  The individual scene event type would provide users with the parameters only associated with the scene event type making it less complicated to determine what properties within SceneEvent were populated.

[MTT-1297](https://jira.unity3d.com/browse/MTT-1297)

## Testing and Documentation
* Includes integration test NetworkSceneManagerTests.SceneEventCallbackNotifications
* This includes XML documentation for:
  * the newly added events.
  * updated/revised some areas for better clarity

## Updated API
- [X] An [api update] was added

NoelStephensUnity committed:
fix

Renamed the SceneEventTypes to match standards and for better clarity.
Adjusted all places that referenced the previous SceneEventTypes with the new SceneEventTypes.

NoelStephensUnity committed:
fix

Provide users with the option to subscribe to each individual SceneEventType notification as opposed to receiving all of them via OnSceneEvent.

Added 8 additional events.  One per each event type in SceneEventData.SceneEventTypes.  Skipped the ReSynchronization event as that might be removed before v1.0.0 depending upon how snapshot and buffered messages are working.

NoelStephensUnity committed:
style

Fixing single white space issue to keep in compliance with standards.

NoelStephensUnity committed:
Merge branch 'develop' into fix/improved-SceneEventTypes-and-new-SceneEventHandlers

becksebenius-unity commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/SceneEventData.cs:
```diff
@@ -15,9 +15,6 @@ public class SceneEventData : IDisposable
-        /// A Server To Client Event (S2C)
-        /// A Client to Server Event (C2S)
         /// </summary>
         public enum SceneEventTypes : byte
```
While you're renaming things, typically Enum type names are singular (so `SceneEventType`)

NoelStephensUnity committed:
refactor

Changed SceneEventTypes to SceneEventType, removed it from SceneEventData and made SceneEventData internal (users don't need to access this).
Did lots of updates on comments for clarity.

NoelStephensUnity committed:
Merge branch 'develop' into fix/improved-SceneEventTypes-and-new-SceneEventHandlers

NoelStephensUnity committed:
fix - merge

fixing some minor merge issues.

becksebenius-unity commented:
Just realized that tools may need an update after this goes in. We're mirroring the SceneEventType names in the tools repository so they'll be out of sync. @josiemessa for awareness. We do have a tech debt to potentially move to strings instead of enum values for this type of thing, but not this release.

NoelStephensUnity committed:
fix and style

Updated the SceneEventTests to reflect the enum changes.
Updated some minor standards issues and minor spelling issues.

NoelStephensUnity committed:
Merge branch 'develop' into fix/improved-SceneEventTypes-and-new-SceneEventHandlers

NoelStephensUnity committed:
refactor

Adding global using ToolsSceneEventType t to point to the tools version of Unity.Multiplayer.Tools.MetricTypes.SceneEventType to keep consistent with naming.

NoelStephensUnity committed:
style

fixing white space issue.

becksebenius-unity reviewed (approved):
Approved. We can follow up on the tools team to make the scene event type names accurate

NoelStephensUnity committed:
test

Added integration test that validates the new delegate events are triggering at the right times with the right information.
Had to add a BaseMultiInstanceTests.m_BypassStartAndWaitForClients bool value that allows you to bypass starting the host and clients as well as the waiting for the clients to connect to test for the client synchronization notifications (i.e. happens during the client connection period of time which already passes by the time the actual unit test starts).

NoelStephensUnity committed:
Merge branch 'develop' into fix/improved-SceneEventTypes-and-new-SceneEventHandlers

NoelStephensUnity committed:
test update

Had left out the scene name from the completed notification validation process.
This adds the scene name and validates that it is correct on both sides (client and server).

NoelStephensUnity committed:
style

typo fix.

NoelStephensUnity committed:
Merge branch 'develop' into fix/improved-SceneEventTypes-and-new-SceneEventHandlers

andrews-unity committed:
Merge branch 'develop' into fix/improved-SceneEventTypes-and-new-SceneEventHandlers
